### ISSUE TITLE
docker network inspect <NETWORK-ID> only returns containers local to current node 

### ISSUE THREAD

**Issue by sfsmithcha** [docker/docker#24186]
**Output of `docker version`:**

```
$ docker version
Client:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
$ docker info
Containers: 7
 Running: 2
 Paused: 0
 Stopped: 5
Images: 7
Server Version: 1.12.0-rc2
Storage Driver: overlay
 Backing Filesystem: xfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host null bridge overlay
Swarm: active
 NodeID: 1120tlfjfl8hb8ej0vl4jf0uk
 IsManager: Yes
 Managers: 1
 Nodes: 3
 CACertHash: sha256:e40f879fdf091f783caae7790db4defc1e1828bc31101029ac18cfb2fdbb56b5
Runtimes: default
Default Runtime: default
Security Options: seccomp
Kernel Version: 3.10.0-327.10.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 991.7 MiB
Name: manager1
ID: VVLS:3SXS:G7IZ:I6MG:DOQ6:VDI5:YWPO:E3XF:ZPMH:RP7C:RBAN:W2SG
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

This is more of an enhancement, but I think in the context of a swarm/cluster, I may want to list the containers participating in the entire network, not just on the current node. And see which nodes the containers are running on.

```
$ docker network inspect mynetwork
[
    {
        "Name": "mynetwork",
        "Id": "8ftib96dxooksdj8b4chprun2",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "75.1.0.0/16",
                    "Gateway": "75.1.0.1"
                }
            ]
        },
        "Internal": false,
        "Containers": {
            "8ec6bc3e3cbb2d22d2c66859b9051c99ab7bd753172a57b89b679b7c80787365": {
                "Name": "frontend.1.5natc1sik5v7i8174h3jzazdo",
                "EndpointID": "cef522ff38f26e9a7284bc273a1f083e560baf297f11050c282ec42688047f34",
                "MacAddress": "02:42:4b:01:00:06",
                "IPv4Address": "75.1.0.6/16",
                "IPv6Address": ""
            },
            "b3455bee301a3f700276ff54bba87c2c3418dbfff6c5c117c46c12c23a027d16": {
                "Name": "redis.1.dsej2jdxc9g2qlvm39m19vn60",
                "EndpointID": "93d447d8d90207aca66e7766e2e5537a2525bb8f529a508c67fb27bfd92dbb7a",
                "MacAddress": "02:42:4b:01:00:08",
                "IPv4Address": "75.1.0.8/16",
                "IPv6Address": ""
            }
        },
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "257"
        },
        "Labels": {}
    }
]
```

**Describe the results you expected:**
See all containers on the network and the node where they're running regardless of the current node.

**Comment by dperny** [comment]
The current implementation of checking services associated with a network is a really expensive operation and kind of a hack, because we don't have an easy way in swarmkit to map network->service. It's definitely possible it just might not be ideal until we can do network->service mapping correctly.

**Comment by DylanSp** [comment]
I'm still seeing this error with Docker 17.12.

Output of `docker version`, from a swarm manager node:
```
Client:
 Version:	17.12.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	c97c6d6
 Built:	Wed Dec 27 20:11:19 2017
 OS/Arch:	linux/amd64

Server:
 Engine:
  Version:	17.12.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	c97c6d6
  Built:	Wed Dec 27 20:09:53 2017
  OS/Arch:	linux/amd64
  Experimental:	false
```

Output of `docker info`, from the same swarm manager:
```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 17.12.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 0
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: c7hjf03p1anudpt98i0solnsk
 Is Manager: true
 ClusterID: pey27mgw739mccti0xs6fnril
 Managers: 3
 Nodes: 6
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 172.24.25.204
 Manager Addresses:
  172.24.25.204:2377
  172.24.25.234:2377
  172.24.25.245:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 89623f28b87a6004d4b785663257362d1658a729
runc version: b2567b37d7b75eb4cf325b77297b140ea686ce8f
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-1022-aws
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.952GiB
Name: <hostname>
ID: F4J4:TQHU:DOST:ECCH:2CFA:IG2C:AKIX:25IQ:4OJN:HT5N:VEPU:J2XJ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

On the swarm manager, `docker network inspect` doesn't show any containers:
```
(manager node):~# docker network inspect us-prod
[
    {
        "Name": "us-prod",
        "Id": "k5s3mc91rxdpl6kdjjw5i2eu5",
        "Created": "0001-01-01T00:00:00Z",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": []
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": null,
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "4097"
        },
        "Labels": null
    }
]
```

But on swarm worker nodes hosting containers attached to that network, `docker network inspect` shows the attached containers:
```
(worker node):~# docker network inspect us-prod
[
    {
        "Name": "us-prod",
        "Id": "k5s3mc91rxdpl6kdjjw5i2eu5",
        "Created": "2018-02-04T15:52:08.6248839Z",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.0.0.0/24",
                    "Gateway": "10.0.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "081feef054d0715241a6578443514602d857cb2d642bdb81c45b0c16e368bc60": {
                "Name": "redis.1.r7je25brg30j5vdzgnw61aqqm",
                "EndpointID": "8cd0f70dda9c9e4d159f899b8ea82aef8481f72e492ff4f7cb267c222ea6d14c",
                "MacAddress": "02:42:0a:00:00:4c",
                "IPv4Address": "10.0.0.76/24",
                "IPv6Address": ""
            },
            "0c4a672c31ac40e04629c3af6f8e0e6c1a318acd56af5290e2f09e64f985b7fa": {
                "Name": "us-prod-sofa.1.dds3a5x5rlsk0wgisg7fhy3nh",
                "EndpointID": "8ec94ced48e186ec38afa0a628075521bcb29cd82f6d6509af2cfca07eeb72bb",
                "MacAddress": "02:42:0a:00:00:15",
                "IPv4Address": "10.0.0.21/24",
                "IPv6Address": ""
            },
            "6d9611ab6cf8ed58f9fea7184858cd744e689964d5e02365930a7ed016d9c59f": {
                "Name": "us-prod-sl.1.w9wdtsxqra97y8oud3pmsp4to",
                "EndpointID": "b39631a8d2e9ef584a433152a744b037b9ea40b01ec8fb4d925249236c50caa1",
                "MacAddress": "02:42:0a:00:00:56",
                "IPv4Address": "10.0.0.86/24",
                "IPv6Address": ""
            },
            "889730f3e202c93bb14fc03f777d5886c47d15f86f1c961f9b0ad6755c358e2a": {
                "Name": "us-prod-soss.1.o1n230hzb977nmd49jmi25tso",
                "EndpointID": "9a6a205bd4204ee3b284868b02126ef0b46246f6e686e499452fb38ee0146559",
                "MacAddress": "02:42:0a:00:00:5d",
                "IPv4Address": "10.0.0.93/24",
                "IPv6Address": ""
            },
            "a651f2cce1bfe3d020af38730e056ca5745429298b6fc63f94c1a0d04ae65b5e": {
                "Name": "memcached.1.y4waq8ejybvlbstbwca7ls1fa",
                "EndpointID": "1fc311b60a5b7e30234fe02ac650446aaabb9fd9ef530c34b1cdf1e97e21e128",
                "MacAddress": "02:42:0a:00:00:4b",
                "IPv4Address": "10.0.0.75/24",
                "IPv6Address": ""
            },
            "bf2efc4f67eef5cb7edb8c746c0e3425fc1afeca42d024e91176afca1add7683": {
                "Name": "us-prod-docs.2.fzar0dl8o0o2qx6aqyb93oy30",
                "EndpointID": "3cd68513e202224f0c5581791edc19c4cd31376a25b68a0185c0d4302ace3ffa",
                "MacAddress": "02:42:0a:00:00:0c",
                "IPv4Address": "10.0.0.12/24",
                "IPv6Address": ""
            },
            "ef2c3e40821b1d45d5566906736b98a44aa734404dc5ff5ca929712afa04ff8b": {
                "Name": "us-prod-soss-datafix.1.0uzfa4uqg8lyj1zsi5tnwgxi4",
                "EndpointID": "ca38756b4d3d570968151678e599437b96e8ca8a0b7efb34e5993c68b530dbbd",
                "MacAddress": "02:42:0a:00:00:1d",
                "IPv4Address": "10.0.0.29/24",
                "IPv6Address": ""
            }
        },
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "4097"
        },
        "Labels": {},
        "Peers": [
            {
                "Name": "2c023f4bcc7f",
                "IP": "172.24.25.200"
            }
        ]
    }
]
```

**Comment by ddebroy** [comment]
If you do `docker network inspect <network-name> -v` you should be able to see the services info as well. The `Host IP` field for each task shows the swarm node running the task.

Example:

```
$ docker network inspect ddrnet1 -v

    {
        "Name": "ddrnet1",
        "Id": "mrzmpv49u8nmrcm4h4lzbk8mb",
        "Created": "2018-02-08T17:35:13.215637022Z",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        .
        .
        .
        "Containers": {
            "0cd71e016244199abfe6fe5f1df8bdd4857e3a2d0b7109705c6e31f0afc1404e": {
                "Name": "nginx1.5.kvm83j4en2aexfv3eihx7ca8g",
                "EndpointID": "c923980c7911cbb3025ab1cf830bd347da553b2294731bbe4376f3d097d00353",
                "MacAddress": "02:42:0a:00:00:0b",
                "IPv4Address": "10.0.0.11/24",
                "IPv6Address": ""
            },
            "7104de5ff971245e50454d5a43323dcc0a432fde45a2455d1a67419af2d8306a": {
                "Name": "nginx1.2.u858twpno4ujla4akzcundjzv",
                "EndpointID": "4e2989db1f131cba004966855b368b0dd03be63b2bbd9ede15f27559e9c89257",
                "MacAddress": "02:42:0a:00:00:08",
                "IPv4Address": "10.0.0.8/24",
                "IPv6Address": ""
            }
        },
        .
        .
        .
        "Services": {
            "nginx1": {
                "VIP": "10.0.0.5",
                "Ports": [],
                "LocalLBIndex": 256,
                "Tasks": [
                    {
                        "Name": "nginx1.2.u858twpno4ujla4akzcundjzv",
                        "EndpointID": "4e2989db1f131cba004966855b368b0dd03be63b2bbd9ede15f27559e9c89257",
                        "EndpointIP": "10.0.0.8",
                        "Info": {
                            "Host IP": "10.0.0.4"
                        }
                    },
                    {
                        "Name": "nginx1.6.j2ui1t27tf1irdghkvn9zbpi8",
                        "EndpointID": "658a4a40684eebbe140a0f28c21666374b4223b3053e763ca16938ec58c5f309",
                        "EndpointIP": "10.0.0.6",
                        "Info": {
                            "Host IP": "10.0.0.6"
                        }
                    },
                    {
                        "Name": "nginx1.1.5dftngubm6mpea1rpk8s8jk1o",
                        "EndpointID": "bf80175b72526c5622ce4a1de1044e29eede0d86a72b66cf6fcdc1887ff408dd",
                        "EndpointIP": "10.0.0.7",
                        "Info": {
                            "Host IP": "10.0.0.5"
                        }
                    },
                    {
                        "Name": "nginx1.5.kvm83j4en2aexfv3eihx7ca8g",
                        "EndpointID": "c923980c7911cbb3025ab1cf830bd347da553b2294731bbe4376f3d097d00353",
                        "EndpointIP": "10.0.0.11",
                        "Info": {
                            "Host IP": "10.0.0.4"
                        }
                    },
                    {
                        "Name": "nginx1.3.xrudl156h5o21lk2wiaacmrmg",
                        "EndpointID": "ef58f08c34ca17f22bcdaa8f93fed767a03fc830fd14d267d7054460bcbf16bf",
                        "EndpointIP": "10.0.0.9",
                        "Info": {
                            "Host IP": "10.0.0.6"
                        }
                    },
                    {
                        "Name": "nginx1.4.kcbv3ooaxyy3uebgp3h6qc9hp",
                        "EndpointID": "3fd1cde738a333ef4669e0288ce29bbfbb01dd3c59a15dd28595b6beb20ccef1",
                        "EndpointIP": "10.0.0.10",
                        "Info": {
                            "Host IP": "10.0.0.5"
                        }
                    }
                ]
            }
        }
    }
```

**Comment by DylanSp** [comment]
@ddebroy Even with -v, it doesn't display the containers or services; the output is the same.

**Comment by thaJeztah** [comment]
not able to reproduce this; on a manger that doesn't have any containers running, I can see the containers that are attached to the network, but are running on other nodes

**Comment by DylanSp** [comment]
Looks like it was an issue due to upgrading from 17.05 (by way of 17.09) in place. I started fresh hosts today with a clean install of 17.12, and I'm able to see containers on other nodes with -v.

**Comment by mcodd** [comment]
I am running 17.12 (clean install from a few weeks ago) and have this same issue.

```
$ sudo docker version
Client:
 Version:       17.12.0-ce
 API version:   1.35
 Go version:    go1.9.2
 Git commit:    c97c6d6
 Built: Wed Dec 27 20:11:19 2017
 OS/Arch:       linux/amd64

Server:
 Engine:
  Version:      17.12.0-ce
  API version:  1.35 (minimum version 1.12)
  Go version:   go1.9.2
  Git commit:   c97c6d6
  Built:        Wed Dec 27 20:09:53 2017
  OS/Arch:      linux/amd64
  Experimental: false
$ 
```

I wonder if it's something that's happening over time and the folks who are unable to reproduce it haven't had enough activity deploying/redeploying things on their swarms to have caused it to happen?  Specifically, against a manager node:

```
$ docker --host tcp://swarm:2376 --tlsverify network inspect nonprod-swarm_uat34_dmz -v
[
    {
        "Name": "nonprod-swarm_uat34_dmz",
        "Id": "4u4t12a5lr3wq8uox9k0sp5fx",
        "Created": "0001-01-01T00:00:00Z",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "",
            "Options": null,
            "Config": null
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": null,
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "4109",
            "encrypted": ""
        },
        "Labels": {
            "com.docker.stack.namespace": "nonprod-swarm"
        }
    }
]
$ 
```

And now directly against a worker node that has a container on that network:

```
[14:25]omni-mgmt-control-0 ~/ansible/omni/roles/simplemon/templates$ docker --host tcp://omni-nonprod-proxyworker-b2:2376 --tlsverify network inspect nonprod-swarm_uat34_dmz -v
[
    {
        "Name": "nonprod-swarm_uat34_dmz",
        "Id": "4u4t12a5lr3wq8uox9k0sp5fx",
        "Created": "2018-02-15T16:02:21.868545032-05:00",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.0.12.0/24",
                    "Gateway": "10.0.12.1"
                }   
            ]   
        },  
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "0bd7d36480e21b4101cb759d6b52b6e61ebba55a4cd7fbc0aed89e3982aa668a": {
                "Name": "uat34_squid.1.n9fyzy5lu2lcrwszgpspos4dr",   
                "EndpointID": "aa0d06b57b6bf850605d28add0e7c3ad47454d90d245259d031811e9b2908067",
                "MacAddress": "02:42:0a:00:0c:2d",
                "IPv4Address": "10.0.12.45/24",
                "IPv6Address": ""
            },
            "1c282d8e4f94443dda97d9fdce9d2b1247275f98ec95b173adcc0ed3ba67e39f": {
                "Name": "nonprod-swarm_postfix.2.vcn9ywb2kxtslk2t16z0mic9v",
                "EndpointID": "29b5efb9748919ff9727e850972efff75a97adf85f945dc4a28a3af299aecf78",
                "MacAddress": "02:42:0a:00:0c:23",
                "IPv4Address": "10.0.12.35/24",
                "IPv6Address": ""
            }
        },
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "4109",
            "encrypted": ""
        },
        "Labels": {
            "com.docker.stack.namespace": "nonprod-swarm"
        },
        "Peers": [
            {
                "Name": "b1cb323c92e7",
                "IP": "172.22.32.10"
            },
            {
                "Name": "b45c4967f9c0",
                "IP": "172.22.35.167"
            },
            {
                "Name": "cc73d7fb1d5f",
                "IP": "172.22.36.161"
            },
            {
                "Name": "81038570033f",
                "IP": "172.22.32.150"
            },
            {
                "Name": "61d5fd0008b6",
                "IP": "172.22.32.248"
            },
            {
                "Name": "52de2c83736d",
                "IP": "172.22.35.60"
            },
            {
                "Name": "a174791a2915",
                "IP": "172.22.32.188"
            }
        ],
        "Services": {
            "nonprod-swarm_metricproxy": {
                "VIP": "10.0.12.32",
                "Ports": [],
                "LocalLBIndex": 415,
                "Tasks": [
                    {
                        "Name": "nonprod-swarm_metricproxy.1.m5d9l3y4finjmwkyg09ewonbl",
                        "EndpointID": "28987df569dac8cceedf41d96ea37719ac92190d2dbe15b86bd56c172e776180",
                        "EndpointIP": "10.0.12.30",
                        "Info": {
                            "Host IP": "172.22.32.10"
                        }
                    },
                    {
                        "Name": "nonprod-swarm_metricproxy.2.uhw30eadknbwg16g25zxanwo4",
                        "EndpointID": "4330da3277963e16ac35e51a26eeaec7c96c4a7cf5679fef7d2fe1c1eb14acab",
                        "EndpointIP": "10.0.12.38",
                        "Info": {
                            "Host IP": "172.22.32.150"
                        }
                    },
                    {
                        "Name": "nonprod-swarm_metricproxy.3.tbulko1i9w9ehgzosxnwdrb64",
                        "EndpointID": "d94140762c5526dbb46ead6701769df771e1334b9a6db3022a6683098d244baf",
                        "EndpointIP": "10.0.12.33",
                        "Info": {
                            "Host IP": "172.22.35.60"
                        }
                    }
                ]
            },
            "nonprod-swarm_postfix": {
                "VIP": "10.0.12.36",
                "Ports": [],
                "LocalLBIndex": 523,
                "Tasks": [
                    {
                        "Name": "nonprod-swarm_postfix.1.j3c8ccubnkz7bjuvdmbxvzp0n",
                        "EndpointID": "adc46e8038c6aaedf5c00b6b6af87afb7140a1db42f6d41c94593e8e3ba6f4a5",
                        "EndpointIP": "10.0.12.37",
                        "Info": {
                            "Host IP": "172.22.32.150"
                        }
                    },
                    {
                        "Name": "nonprod-swarm_postfix.2.vcn9ywb2kxtslk2t16z0mic9v",
                        "EndpointID": "29b5efb9748919ff9727e850972efff75a97adf85f945dc4a28a3af299aecf78",
                        "EndpointIP": "10.0.12.35",
                        "Info": {
                            "Host IP": "172.22.36.161"
                        }
                    }
                ]
            },
            "nonprod-swarm_xray-daemon": {
                "VIP": "10.0.12.28",
                "Ports": [],
                "LocalLBIndex": 350,
                "Tasks": [
                    {
                        "Name": "nonprod-swarm_xray-daemon.1.xpu1iy1uj9lr86w2338rkl053",
                        "EndpointID": "2ec6834d5cf2c818071ed454a105aa4f7df1fa03ffbc44a630385be8f958052b",
                        "EndpointIP": "10.0.12.29",
                        "Info": {
                            "Host IP": "172.22.35.167"
                        }
                    },
                    {
                        "Name": "nonprod-swarm_xray-daemon.2.1zh89wst7n71clywqoy65i2bb",
                        "EndpointID": "2a743b6cae97ac4eb773cb53eab9d4d7aa81ce68df2974dc6621dff78d294b52",
                        "EndpointIP": "10.0.12.34",
                        "Info": {
                            "Host IP": "172.22.32.150"
                        }
                    },
                    {
                        "Name": "nonprod-swarm_xray-daemon.3.q16cfmir4fjnuq3gtz8ohanjp",
                        "EndpointID": "2c430cbb5cc1c41b7692f19589f8d2148ccd4228f7417ca973fd973b4bccf130",
                        "EndpointIP": "10.0.12.31",
                        "Info": {
                            "Host IP": "172.22.32.10"
                        }
                    }
                ]
            },
            "uat34_notepasser": {
                "VIP": "10.0.12.42",
                "Ports": [],
                "LocalLBIndex": 846,
                "Tasks": [
                    {
                        "Name": "uat34_notepasser.1.cl1k2wuihn12l16vcf57dr1yq",
                        "EndpointID": "99b95f9c4c261cef3137067a914e39b83cb30449446c53c860bb65812acee56f",
                        "EndpointIP": "10.0.12.43",
                        "Info": {
                            "Host IP": "172.22.32.248"
                        }
                    }
                ]
            },
            "uat34_squid": {
                "VIP": "10.0.12.44",
                "Ports": [],
                "LocalLBIndex": 825,
                "Tasks": [
                    {
                        "Name": "uat34_squid.1.n9fyzy5lu2lcrwszgpspos4dr",
                        "EndpointID": "aa0d06b57b6bf850605d28add0e7c3ad47454d90d245259d031811e9b2908067",
                        "EndpointIP": "10.0.12.45",
                        "Info": {
                            "Host IP": "172.22.36.161"
                        }
                    }
                ]
            }
        }
    }
]
$ 
```

**Comment by mcodd** [comment]
Doing a bit more digging here, this is totally related to whether the manager node I'm talking to has a container on the network in question, similar to what was previously reported here.  I had missed this pattern in my previous testing and take back what I said about this maybe happening only after the swarm has been up for some time.  I brought up a fresh swarm and can reproduce it.  Once I update a service running in a manager to attach to a network that didn't used to have any containers attached to it on that manager, it shows up in a `docker network inspect` (with our without the -v) against that manager.

What I did notice was that the "fresh install" comment I made earlier isn't 100% true - we have an AMI that includes docker 17.09 and after using terraform to instantiate the swarm nodes, but prior to bringing up the swarm, we have an Ansible playbook running that's upgrading that to 17.12 (then we reboot all the nodes prior to actually creating and joining the new swarm).  I find it really hard to believe that this could cause the issue being discussed here but because @DylanSp noted that fixed his issue I'm going to update our AMI with one that has never had any traces of anything on it besides 17.12 and see where I end up...

**Comment by thaJeztah** [comment]
@mcodd on 17.12.0 and older, using a _partial_ network-id for inspect may not show the detailed information. That issue was fixed in docker 17.12.1 (see https://github.com/moby/moby/pull/35989 / https://github.com/docker/docker-ce/pull/423)

**Comment by mcodd** [comment]
In my case there is no difference when using a completely fresh install on 17.12... networks that do not have any containers running on them on the manager i query with `docker network inspect` do not show any services running on that network.  Again as an example, similar to what i showed last time, this is against a swarm manager:

```
$ sudo docker version
Client:
 Version:       17.12.0-ce
 API version:   1.35
 Go version:    go1.9.2
 Git commit:    c97c6d6
 Built: Wed Dec 27 20:11:19 2017
 OS/Arch:       linux/amd64

Server:
 Engine:
  Version:      17.12.0-ce
  API version:  1.35 (minimum version 1.12)
  Go version:   go1.9.2
  Git commit:   c97c6d6
  Built:        Wed Dec 27 20:09:53 2017
  OS/Arch:      linux/amd64
  Experimental: false
$ 
```

```
$ docker --tlsverify --host tcp://omni-sandbox-manager-g1:2376 network inspect sandbox-swarm_dmz -v
[
    {
        "Name": "sandbox-swarm_dmz",
        "Id": "bfw6o1hnmh7ocjwmzbngqjwe4",
        "Created": "0001-01-01T00:00:00Z",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "",
            "Options": null,
            "Config": null
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": null,
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "4098",
            "encrypted": ""
        },
        "Labels": {
            "com.docker.stack.namespace": "sandbox-swarm"
        }
    }
]
$ 
```

And now against a worker node that actually has a container running on the sandbox-swarm_dmz network:

```
$ docker --tlsverify --host tcp://omni-sandbox-proxyworker-g1:2376 network inspect sandbox-swarm_dmz -v
[
    {
        "Name": "sandbox-swarm_dmz",
        "Id": "bfw6o1hnmh7ocjwmzbngqjwe4",
        "Created": "2018-03-16T20:12:18.698870148-04:00",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.0.1.0/24",
                    "Gateway": "10.0.1.1"
                }   
            ]   
        },  
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },  
        "ConfigOnly": false,
        "Containers": {
            "053c214c177ba0b34318ed0165a3c9f57d23ce972f484b10003c9f08d675c413": {
                "Name": "sandbox-swarm_postfix.2.lq2o1z8onj2zilo3iu26ysbt3",   
                "EndpointID": "d0a90c213d2d403453e4f3c126cb7af4899c5551c8c2c042d865bfd14f2c863b",
                "MacAddress": "02:42:0a:00:01:0d",
                "IPv4Address": "10.0.1.13/24",
                "IPv6Address": ""
            },  
[[[ TRUNCATING LONG OUTPUT HERE ]]]
            "sandbox-swarm_xray-daemon": {
                "VIP": "10.0.1.18",
                "Ports": [],
                "LocalLBIndex": 270,
                "Tasks": [
                    {   
                        "Name": "sandbox-swarm_xray-daemon.3.je5aaigrh2fenvoada1jwn8n1",
                        "EndpointID": "6c099cb5aa0c23ef81118789c74cde8ec7bb31fe9dc07b025170c96202059494",
                        "EndpointIP": "10.0.1.21",
                        "Info": {
                            "Host IP": "172.24.32.222"
                        }
                    },
                    {   
                        "Name": "sandbox-swarm_xray-daemon.1.nvzu8hil9ew3ijzel2zsxmliw",
                        "EndpointID": "0a7988ddcfdeb1e39c3ac973208b14673562d1990ea8662720f18eb4b28c50ac",
                        "EndpointIP": "10.0.1.19",
                        "Info": {
                            "Host IP": "172.24.32.222"
                        }
                    },
                    {   
                        "Name": "sandbox-swarm_xray-daemon.2.nviw1rdr02370xayylg4stl6c",
                        "EndpointID": "fb704fd7b43049b9beddebfdeb8801e64c8b334f6df8b8113c69a1e6302b0c03",
                        "EndpointIP": "10.0.1.20",
                        "Info": {
                            "Host IP": "172.24.35.245"
                        }
                    }
                ]
            }
        }
    }
]
```

So, I'm curious about the people who are seeing `docker network inspect $network` output that includes services running on $network in a situation where no containers running on the manager queried live on that particular network.  I guess the question is, what's the difference between my configuration and those folks'? I definitely don't see that running Docker 17.09 vs 17.12 makes a difference at all in my experiments.

matt

**Comment by mcodd** [comment]
@thaJeztah - thanks for the response!  I hadn't seen it before I typed up mine... tried replacing the name i was using with its full ID, but no improvement.  Note that it's not just extra/verbose details i'm missing, even the "Created" time is bogus, for example:

```
$ docker --tlsverify --host tcp://omni-sandbox-manager-g1:2376 network inspect bfw6o1hnmh7ocjwmzbngqjwe4 -v
[
    {
        "Name": "sandbox-swarm_dmz",
        "Id": "bfw6o1hnmh7ocjwmzbngqjwe4",
        "Created": "0001-01-01T00:00:00Z",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "",
            "Options": null,
            "Config": null
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": null,
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "4098",
            "encrypted": ""
        },
        "Labels": {
            "com.docker.stack.namespace": "sandbox-swarm"
        }
    }
]
$ 
```

**Comment by thaJeztah** [comment]
Perhaps it’s best to start a fresh issue (given that the initial issue here was an enhancement request); could you open a new ticket with the details you provided?

**Comment by mcodd** [comment]
Yep, let me first try 17.12.1 anyway and if that doesn't solve things i'll provide the details against that version.

**Comment by siddjain** [comment]
Hi, we are seeing this exact same behavior using docker 18.06. I read the comments above and bit confused. So what's the upshot of all this? Is it expected behavior of docker that network inspect only lists containers hosted on the node from which the command is executed?

**Comment by persunde** [comment]
I am having this issue on Docker Swarm: `Docker version 20.10.17, build 100c701`

`docker network inspect <network-name> [-v]` only displays containers that are running on that host. I assume it should show ALL containers that are part of the network running on any host/node.

**Comment by thaJeztah** [comment]
@persunde make sure to look at the `tasks` under `Services`, not the `containers` list; the `tasks` represent services connected to the network (which may not be local); `containers` will only contain containers on the current node.

For example;


```bash
docker network inspect mynet -v
[
    {
        "Name": "mynet",
        "Id": "xr1vyuixwvbzj1607r63tr7bk",
        "Created": "2022-08-19T15:53:45.051778821Z",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.0.0.0/24",
                    "Gateway": "10.0.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "c391b6ead0925762132cec2114e6c564b56f88b5fe582f92032e982aa9fdbe75": {
                "Name": "testing.3.ybbtitbu2ih4dsnh67go7afcw",
                "EndpointID": "774a7e17dfaeb417c9393ecfa6160e9b76fa480e69529a12dac6967623db65b8",
                "MacAddress": "02:42:0a:00:00:04",
                "IPv4Address": "10.0.0.4/24",
                "IPv6Address": ""
            },
            "lb-mynet": {
                "Name": "mynet-endpoint",
                "EndpointID": "372d846f8960d7a3005f1b3a4661fcdce9b691a6eb653cb82aaf5af3d62a2405",
                "MacAddress": "02:42:0a:00:00:07",
                "IPv4Address": "10.0.0.7/24",
                "IPv6Address": ""
            }
        },
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "4097"
        },
        "Labels": {},
        "Peers": [
            {
                "Name": "5b12e95f8b48",
                "IP": "10.133.83.180"
            },
            {
                "Name": "f4af3629c59a",
                "IP": "10.133.77.81"
            },
            {
                "Name": "519d1bbb8317",
                "IP": "10.133.84.218"
            }
        ],
        "Services": {
            "testing": {
                "VIP": "10.0.0.2",
                "Ports": [],
                "LocalLBIndex": 256,
                "Tasks": [
                    {
                        "Name": "testing.3.ybbtitbu2ih4dsnh67go7afcw",
                        "EndpointID": "774a7e17dfaeb417c9393ecfa6160e9b76fa480e69529a12dac6967623db65b8",
                        "EndpointIP": "10.0.0.4",
                        "Info": {
                            "Host IP": "10.133.83.180"
                        }
                    },
                    {
                        "Name": "testing.1.siesp65lugjxyitam25ixr9do",
                        "EndpointID": "949ddfc77c0d5a15fa5e719d8f43c4625ade2e95e2aa29f6a88d9fd04edddfd4",
                        "EndpointIP": "10.0.0.5",
                        "Info": {
                            "Host IP": "10.133.84.218"
                        }
                    },
                    {
                        "Name": "testing.2.7avts9h56covj8419686ztfo6",
                        "EndpointID": "a886dda37b77ec47396aa52df45fde34bc4a0fd6bfda90c54777ba91c50664af",
                        "EndpointIP": "10.0.0.3",
                        "Info": {
                            "Host IP": "10.133.77.81"
                        }
                    }
                ]
            }
        }
    }
]
```

**Comment by dr-yd** [comment]
This issue is still present in Docker version 27.1.2, build d01f264. For Swarm overlay networks that have containers running only on worker nodes, running `docker network inspect -v` on the manager contains neither services nor peers. Sample:

Manager:
```json
[
    {
        "Name": "foo-bar_backend",
        "Id": "j2dugom70x30d7w5qd8w0v1d1",
        "Created": "2024-08-25T16:39:28.535335018Z",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.24.24.0/22",
                    "Gateway": "10.24.24.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": null,
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "4120"
        },
        "Labels": {
            "com.docker.stack.namespace": "foo-bar"
        }
    }
]
```

Worker:
```json
[
    {
        "Name": "foo-bar_backend",
        "Id": "j2dugom70x30d7w5qd8w0v1d1",
        "Created": "2024-09-03T18:18:40.82121228+02:00",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.24.24.0/22",
                    "Gateway": "10.24.24.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "9c7160cf9790eef22f1e239dabe448f68624e9a5bc89d4b5169e6c8b88f3bcf5": {
                "Name": "foo-bar_baz.1.qccv443xwobbyhbtrfio7j5iv",
                "EndpointID": "0e9c9ec6d27b537ee7255d456d8fdbf0bc798cb75d0b6e3a833335ad689b10bc",
                "MacAddress": "02:42:0a:18:18:1c",
                "IPv4Address": "10.24.24.28/22",
                "IPv6Address": ""
           [...]
     }
]
```

As soon as a service running on the manager takes part in the network, containers become visible, but that's of course not very useful.
